hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/MetricsService.java

```diff
@@ -147,13 +148,21 @@ MetricsConfig getConfig() {
 
     // visible for testing
     void collectMetrics() {
+        PublisherMetricsCollector publisherCollector = new PublisherMetricsCollector(publisherArr());
         collectMetrics(publisherCollector);
+        publisherCollector.publishCollectedMetrics();
+    }
+
+    private MetricsPublisher[] publisherArr() {
+        Object[] objPublishers = publishers.toArray();
```

Comment:
I think you can just to the `publishers.toArray(T[] t)` overload with 0 size array to avoid cloning here..

Reply:
@cang I think that can be safe only if we do it in a `synchronized` block (or with some other concurrency control), since we need to know the size of the `t` array upfront. Without that, between taking the size and doing the actual copy one can concurrently register a publisher, which leads to an error. I thought cloning here is better than synchronizing.

Reply:
COWList internally does this, without requiring synchronization.

Reply:
Uh, right :) Replaced.